### PR TITLE
[Feature] - Mailer Presence Checker

### PR DESCRIPTION
**Pull request by AyakorK**
This Pull Request will send a mail at 18h00 (6:00PM) to send to the supervisor of each class of active courses the list of absences that were listed at this day on the two periods (morning and afternoon [symbolized by "arrival" and "departure"])

This might not be the most optimized way of doing it but it still works pretty well and needs to be tested in production, In addition of that we're going to see some changes, like the addition of a checker that we're not in weekend to forbid some actions to occur on those days and also some corrections essentially technical

### CODE REVIEW

File: back-end/src/base/calls/calls.controller.ts [hunk 1]
```diff
@@ -30,7 +30,6 @@ export class CallsController {
 		return res.status(201).json({ status: 'ok', qrcode: qrcode });
 	}
 	@Get('/presence/:jwt')
-	@UseGuards(JwtAuthGuard)
```
Comment (Maghwyn) [inline back-end/src/base/calls/calls.controller.ts:33]: The reason ?

File: back-end/src/base/calls/calls.repository.ts [hunk 2]
```diff
@@ -17,10 +26,176 @@ export class CallsRepository {
...
+			.find({
+				$and: [{ periodStart: { $lte: actualDate } }, { periodEnd: { $gte: actualDate } }],
+			})
+			.toArray();
+	}
+
+	async getClassSupervisor(courseId: InferIdType<Course>) {
+		const courseObjectId = new ObjectId(courseId);
+		const course = await this.courses.findOne({ _id: courseObjectId });
+		// @ts-ignore (Mandatory because of typo issue)
```
Comment (Maghwyn) [inline back-end/src/base/calls/calls.repository.ts]: Your id should be an objectId in the database btw.
And i just verified, it is.
So you don't need to convert it when it's already one.

File: back-end/src/base/calls/calls.repository.ts [hunk 3]
```diff
@@ -17,10 +26,176 @@ export class CallsRepository {
...
+		const classObjectId = new ObjectId(course.classId);
+		const classroom = await this.db.collection('classes').findOne({ _id: classObjectId });
+		return await this.db.collection('users').findOne({ _id: new ObjectId(classroom.pedago) });
+	}
+
+	async getClass(courseId: InferIdType<Course>) {
+		const courseObjectId = new ObjectId(courseId);
+		const course = await this.courses.findOne({ _id: courseObjectId });
+		// @ts-ignore (Mandatory because of typo issue)
```
Comment (Maghwyn) [inline back-end/src/base/calls/calls.repository.ts]: Same here

File: back-end/src/base/calls/calls.service.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ export class CallsService {
 	}
 
 	async generateUrl(jwt: string) {
-		return `https://1b68-2a01-cb00-e91-b600-2d60-c9db-21f5-24f.eu.ngrok.io/calls/presence/` + jwt;
+		return `https://backend-codingtools.loule.me/calls/presence/` + jwt;
```
Comment (Maghwyn) [inline back-end/src/base/calls/calls.service.ts:39]: Add a comment TODO to specify that it needs to be under a dotenv variable for production.

File: back-end/src/cron/cron.service.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import { AbsencesParams } from '@/base/calls/interfaces/calls.interface';
+import { MaijetTemplate } from 'src/common/providers/interfaces/events.interface';
+
+@Injectable()
+export class CronService {
+	constructor(private readonly callsRepository: CallsRepository) {}
+
+	@Cron('0 18 * * *')
+	async handleCron() {
+		if (this.isWeekend()) {
```
Comment (Maghwyn) [inline back-end/src/cron/cron.service.ts]: Pretty sure you can manage that directly on the cron schedule.
@Cron('0 18 * * 1-5')

**Review by Maghwyn — CHANGES_REQUESTED** [review]

File: back-end/src/cron/cron.service.ts (hunk 5 above)
Comment (Maghwyn) [reply back-end/src/cron/cron.service.ts]: [Cron guru](https://crontab.guru/#0_18_*_*_1-5)

File: back-end/src/base/calls/calls.controller.ts (hunk 1 above)
Comment (AyakorK) [reply back-end/src/base/calls/calls.controller.ts:33]: You're using your phone, as you're not logged in you cannot interact with this route if you keep the @UseGuards.

File: back-end/src/base/calls/calls.repository.ts (hunk 2 above)
Comment (AyakorK) [reply back-end/src/base/calls/calls.repository.ts]: Already fixed

File: back-end/src/base/calls/calls.repository.ts (hunk 3 above)
Comment (AyakorK) [reply back-end/src/base/calls/calls.repository.ts]: Same

File: back-end/src/base/calls/calls.service.ts (hunk 4 above)
Comment (AyakorK) [reply back-end/src/base/calls/calls.service.ts:39]: Will be done

File: back-end/src/cron/cron.service.ts (hunk 5 above)
Comment (AyakorK) [reply back-end/src/cron/cron.service.ts]: Nice catch ! That's my bad didn't think about this

**Review by RuKaSuu — APPROVED** [review]

File: back-end/src/base/calls/calls.repository.ts (hunk 2 above)
Comment (Maghwyn) [reply back-end/src/base/calls/calls.repository.ts]: Okay, why is the ts-ignore still there ?

**Review by Maghwyn — APPROVED** [review]

File: back-end/src/base/calls/calls.repository.ts (hunk 2 above)
Comment (Maghwyn) [reply back-end/src/base/calls/calls.repository.ts]: Nevermind i checked.
You have a missmatch of data from what's in the database courses collection and what's in the interface of your courses collection.

Comment (AyakorK) [reply back-end/src/base/calls/calls.repository.ts]: Okay it's fixed thanks for the advice

**Review by Maghwyn — APPROVED** [review]

**Pull request merged**